diff --git a/hw3/main.c b/hw3/main.c
index 3861fe6..8827905 100644
--- a/hw3/main.c
+++ b/hw3/main.c
@@ -448,6 +448,7 @@ void __init parse_early_param(void)
 {
 	static __initdata int done = 0;
 	static __initdata char tmp_cmdline[COMMAND_LINE_SIZE];
+	printk(KERN_INFO "Kai-Chen_Hsieh: parse_early_param(): setup_arch() code calls this early on \n");
 
 	if (done)
 		return;
@@ -491,6 +492,7 @@ static void __init mm_init(void)
 	 * page_cgroup requires contiguous pages,
 	 * bigger than MAX_ORDER unless SPARSEMEM.
 	 */
+	printk(KERN_INFO "Kai Chen, Hsieh: mm_init(): calls mem_init() and kmem_cache_init() \n");
 	page_cgroup_init_flatmem();
 	mem_init();
 	kmem_cache_init();
diff --git a/hw3/page_alloc.c b/hw3/page_alloc.c
index 9ddea02..e4588a3 100644
--- a/hw3/page_alloc.c
+++ b/hw3/page_alloc.c
@@ -5569,6 +5569,7 @@ static int page_alloc_cpu_notify(struct notifier_block *self,
 
 void __init page_alloc_init(void)
 {
+	printk(KERN_INFO "Kai-Chen_Hsieh: page_alloc_init(): this function calls hotcpu_notifier() \n");
 	hotcpu_notifier(page_alloc_cpu_notify, 0);
 }
 
diff --git a/hw3/setup.c b/hw3/setup.c
index b1077b8..72d77b8 100644
--- a/hw3/setup.c
+++ b/hw3/setup.c
@@ -973,6 +973,7 @@ void __init setup_arch(char **cmdline_p)
 
 	strlcpy(command_line, boot_command_line, COMMAND_LINE_SIZE);
 	*cmdline_p = command_line;
+	printk(KERN_INFO "Kai-Chen_Hsieh: setup_arch(): command_line = %s \n", *cmdline_p);
 
 	/*
 	 * x86_configure_nx() is called before parse_early_param() to detect
diff --git a/hw3/traps.c b/hw3/traps.c
index 871f43a..b4ae257 100644
--- a/hw3/traps.c
+++ b/hw3/traps.c
@@ -801,6 +801,7 @@ void __init trap_init(void)
 	early_iounmap(p, 4);
 #endif
 
+	printk(KERN_INFO "Kai Chen, Hsieh: trap_init(): initialize traps \n");
 	set_intr_gate(X86_TRAP_DE, divide_error);
 	set_intr_gate_ist(X86_TRAP_NMI, &nmi, NMI_STACK);
 	/* int4 can be called from all */
